File: tests/php/static_closures/partial_application_via_closure.php
<?php
// vybe-test: php/static_closures/partial_application_via_closure
// origin: languages/php/tests/php/test_static_closures.rs

function __vybe_check($got, $want) {
    // Match the Rust harness's normalisation: strip \r, then drop trailing
    // newlines (it split on "\n" and popped empty trailing elements).
    $got = str_replace("\r", "", $got);
    $got = rtrim($got, "\n");
    if ($got !== $want) {
        echo "FAIL: want [" . $want . "] got [" . $got . "]\n";
        throw new Exception("assertion failed");
    }
    // Replay the program's own output so running the file by hand still
    // behaves like the program it was extracted from.
    echo $got;
    if ($got !== "") {
        echo "\n";
    }
}

ob_start();

function partial(callable $fn, mixed ...$partial): Closure {
    return function() use ($fn, $partial) {
        $args = array_merge($partial, func_get_args());
        return $fn(...$args);
    };
}
$add = fn(int $a, int $b): int => $a + $b;
$add5 = partial($add, 5);
echo $add5(3);

__vybe_check(ob_get_clean(), "8");

File: tests/lua/closures/partial_application_via_closure.lua
-- vybe-test: lua/closures/partial_application_via_closure
-- origin: languages/lua/tests/lua/test_closures.rs

local __w1 = "8,15"
local __i = 0

local function partial(f, a)
  return function(b) return f(a, b) end
end
local add = function(a, b) return a + b end
local add5 = partial(add, 5)
do local __t = tostring(add5(3) .. ',' .. add5(10)); __i = __i + 1
  if __i == 1 and __t ~= __w1 then error("FAIL: want [" .. __w1 .. "] got [" .. __t .. "]") end end

if __i == 0 then error("FAIL: no output, wanted [" .. __w1 .. "]") end
